### ISSUE TITLE
Unexpected error when using config file inheritance with project references

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#40720]
**TypeScript Version:**  4.0.2

**Search Terms:** project references composite configuration file inheritance include files

**Code**

Reduced test case: https://github.com/OliverJAsh/ts-project-references-config-inheritance-bug. Below I have inlined the code from this repository.

Project structure:

```
├── shared
│   ├── index.ts
│   ├── tsconfig-base.json
│   ├── tsconfig.json (extends `shared/tsconfig-base.json`)
│   └── typings-base
│       └── globals.d.ts
├── webpack
│   ├── index.ts
│   └── tsconfig.json (extends `shared/tsconfig-base.json`)
└── tsconfig.json (references projects `shared/tsconfig.json` and `webpack/tsconfig.json`)
```

Contents:

`shared/index.ts`:
```ts
export const a: Unrestricted = 1;
```

`shared/tsconfig-base.json`:
```json
{
  "include": ["./typings-base/"]
}
```

`shared/tsconfig.json`:
```json
{
  "extends": "./tsconfig-base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "../target-tsc-build/",
    "rootDir": ".."
  },
  "files": ["./index.ts"]
}
```

`shared/typings-base/globals.d.ts`:
```ts
type Unrestricted = any;
```

`webpack/index.ts`:
```ts
export const b: Unrestricted = 1;
```

`webpack/tsconfig.json`:
```json
{
  "extends": "../shared/tsconfig-base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "../target-tsc-build/",
    "rootDir": ".."
  },
  "files": ["./index.ts"],
  "references": [
    {
      "path": "../shared/tsconfig.json"
    }
  ]
}
```

`tsconfig.json`:
```json
{
  "references": [
    {
      "path": "./shared/tsconfig.json"
    },
    {
      "path": "./webpack/tsconfig.json"
    }
  ],
  "files": []
}
```

**Expected behavior:**

```bash
# No errors ✅
tsc --build webpack/tsconfig.json --force --verbose

# No errors ✅
tsc --build tsconfig.json --force --verbose
```

**Actual behavior:**

```bash
# No errors ✅
tsc --build webpack/tsconfig.json --force --verbose

# Errors ❌
tsc --build tsconfig.json --force --verbose
```

The unexpected error:

```
[16:49:43] Projects in this build:
    * shared/tsconfig.json
    * webpack/tsconfig.json
    * tsconfig.json

[16:49:43] Project 'shared/tsconfig.json' is up to date because newest input 'shared/typings-base/globals.d.ts' is older than oldest output 'target-tsc-build/shared/index.js'

[16:49:43] Building project '/Users/oliverash/Development/ts-project-references-config-inheritance-bug/shared/tsconfig.json'...

[16:49:44] Project 'webpack/tsconfig.json' is out of date because oldest output 'target-tsc-build/webpack/index.js' is older than newest input 'shared/tsconfig.json'

[16:49:44] Building project '/Users/oliverash/Development/ts-project-references-config-inheritance-bug/webpack/tsconfig.json'...

webpack/index.ts:1:17 - error TS2304: Cannot find name 'Unrestricted'.

1 export const b: Unrestricted = 1;
                  ~~~~~~~~~~~~


Found 1 error.
```

**Playground Link:** 

**Related Issues:**

**Comment by OliverJAsh** [comment]
FWIW it would be useful to have a flag like `--showConfig` specifically for `--build`, to better understand issues like this. This way, we could see what the resolved config will be for referenced projects.

**Comment by OliverJAsh** [comment]
This might not be related but I just ran into another similar case where again I'm seeing inconsistent errors when I run `tsc --build` with a project directly versus when run I run it with a project that references the aforementioned project.

https://github.com/OliverJAsh/ts-project-references-config-inheritance-bug/tree/problem-2

In this case, I'm expecting an error but not getting any.

```bash
# Expected error and got one ✅
tsc --build tsconfig.json --clean && \
    tsc --build tsconfig.json --verbose
```

As expected, this errors with:

```
error TS2318: Cannot find global type 'Array'.

error TS2318: Cannot find global type 'Boolean'.

error TS2318: Cannot find global type 'Function'.

error TS2318: Cannot find global type 'IArguments'.

error TS2318: Cannot find global type 'Number'.

error TS2318: Cannot find global type 'Object'.

error TS2318: Cannot find global type 'RegExp'.

error TS2318: Cannot find global type 'String'.


Found 8 errors.
```

However, if I run use the "solution" project instead, I don't get any errors:

```bash
# Expected error but got none ❌
tsc --build service-worker/tsconfig.json --clean && \
    tsc --build service-worker/tsconfig.json --verbose
```
